src/yb/yql/cql/cqlserver/cql_service.cc

```diff
@@ -229,6 +236,40 @@ void CQLServiceImpl::DeletePreparedStatement(const shared_ptr<const CQLStatement
           << ", memory usage = " << prepared_stmts_mem_tracker_->consumption();
 }
 
+bool CQLServiceImpl::CheckPassword(const std::string plain, const std::string expected_bcrypt_hash) {
+  if (FLAGS_password_hash_cache_size <= 0) {
+    return util::bcrypt_checkpw(plain.c_str(), expected_bcrypt_hash.c_str()) == 0;
+  }
+
+  // FIXME DO NOT MERGE: hash the plaintext password so we're not keeping
+  // multiple plaintext passwords in memory
+
+  std::string sha_hash = plain;
+  // bcrypt can generate multiple hashes from a single key, since a salt is
+  // randomly generated each time a password is set. Using a compound key allows
+  // the same plaintext to be associated with different hashes.
+  std::string key = sha_hash + ":" + expected_bcrypt_hash;
+
+  {
+    std::lock_guard<std::mutex> guard(password_cache_mutex_);
+    auto entry = password_cache_.get(key);
+    if (entry) {
+      return true;
```

Comment:
I have actually been thinking about this, and even with the current approach, it should still work as a negative cache. For example, let's say that r1/p1 is the right username password, and an attempt is made with r1/p2, so we would just insert p2:bcrypt(p1), false into the LRU. Next time another attempt is made to connect with the same combo r1/p2, we would return false.

Let's say that the password does change to p2, then next time a connection attempt is made with r1/p2, we would try to find the key p2:bcrypt(p2) (since bcrypt(p2) is what the master will return), and we won't find it, so we will insert a new entry into the LRU and everything works as expected

Reply:
I'm not sure of the value of this, since most automated systems should have the correct password, and will immediately error out if it is not correct. Also, this would push the correct password entries out of the cache, which could hypothetically be used to slow down connection creation if done maliciously. This wouldn't be a full DoS attack, since it would just regress back to previous behavior, but it's still worth considering.